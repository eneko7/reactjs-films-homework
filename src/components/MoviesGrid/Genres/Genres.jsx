import React from 'react';
import { Link } from 'react-router-dom';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import style from './Genres.scss';

class Genres extends React.Component {
  constructor(props) {
    super(props);
    const { title } = this.props;
    this.state = {
      isOpen: false,
      headerTitle: title,
    };
    this.openGenresClick = this.openGenresClick.bind(this);
    this.changeTitle = this.changeTitle.bind(this);
  }

  componentDidMount() {
    const { location: { search, pathname }, fetchGenres } = this.props;
    const parsed = queryString.parse(search);
    const { genreId, genreName } = parsed;
    fetchGenres();
    if (pathname === '/genres') {
      this.changeTitle(genreName, genreId);
    }
  }

  openGenresClick() {
    this.setState(prevState => ({
      isOpen: !prevState.isOpen,
    }));
  }

  changeTitle(el, id) {
    this.setState(() => ({
      headerTitle: el,
      isOpen: false,
    }));
    const { fetchFilmsByGenre, fetchChange } = this.props;
    fetchFilmsByGenre(id);
    fetchChange('Genres');
  }

  render() {
    const { genres } = this.props;
    const genresCol = genres.map(el => (
      <Link key={el.id} to={`/genres?genreName=${el.name}&genreId=${el.id}`} style={{ textDecoration: 'none' }}>
        <button type="button" id={el.name} className={`${style.moviesGrid_categories_item_block_button}`} elem={el.name} onClick={this.changeTitle.bind(null, el.name, el.id)}>
          {el.name}
        </button>
      </Link>
    ));
    const { isOpen, headerTitle } = this.state;
    const active = isOpen ? `${style.active}` : '';
    const { activeCategory } = this.props;
    const genresActive = activeCategory === 'Genres' ? { borderBottom: '4px solid' } : { borderBottom: 'none' };
    return (
      <div>
        <button type="button" style={genresActive} className={`${style.moviesGrid_categories_item_button}`} onClick={this.openGenresClick} data-filter="Genres" id="Genres">
          <span>{headerTitle}</span>
          <span className={`${active} ${style.moviesGrid_categories_item_span_arrow}`} />
        </button>
        <div className={`${active} ${style.moviesGrid_categories_item_block}`} id="genres_list">
          {genresCol}
        </div>
      </div>
    );
  }
}

Genres.propTypes = {
  fetchGenres: PropTypes.func.isRequired,
  fetchChange: PropTypes.func.isRequired,
  fetchFilmsByGenre: PropTypes.func.isRequired,
  genres: PropTypes.arrayOf(PropTypes.object).isRequired,
  title: PropTypes.string.isRequired,
  activeCategory: PropTypes.string.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string,
    search: PropTypes.string,
  }).isRequired,
};

export default Genres;
