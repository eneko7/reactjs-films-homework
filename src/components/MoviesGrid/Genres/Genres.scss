.moviesGrid_categories_item_button {
  cursor: pointer;
  font-size: 25px;
  line-height: 30px;
  color: #0aaee4;
  font-family: Helvetica, Arial, Verdana, Tahoma, sans-serif;
  margin-right: 30px;
  padding-bottom: 5px;
  text-align: center;
  display: block;
  border: none;
  background: none;

  .moviesGrid_categories_item_span_arrow {
    background-image: url(../../../images/angle-arrow-down.png);
    background-repeat: no-repeat;
    width: 13px;
    height: 13px;
    display: inline-block;
    background-size: contain;
    background-position: center;
    margin-left: 10px;
  }

  .moviesGrid_categories_item_span_arrow.active {
    background-image: url(../../../images/up-arrow.png);
  }
}

.moviesGrid_categories_item_button:focus {
  outline: none;
}

.moviesGrid_categories_background_wraper {
  display: none;
}

.moviesGrid_categories_background_wraper.active {
  display: block;
  position: absolute;
  z-index: 5;
  opacity: 1;
  width: 100vw;
  height: 200vh;
  background: none;
  border: none;
  top: -95vh;
  left: -8.3vw;
}

.moviesGrid_categories_background_wraper.active:focus {
  outline: none;
}

.moviesGrid_categories_item_block {
  list-style-type: none;
  display: none;

  .moviesGrid_categories_item_block_button {
    width: 100%;
    background: none;
    color: #0aaee4;
    border: none;
    cursor: pointer;
    text-align: left;
    font-size: 18px;
    margin-top: 10px;
  }

  .moviesGrid_categories_item_block_button:hover {
    background: #0aaee4;
    color: #fff;
  }

  .moviesGrid_categories_item_block_button:focus {
    outline: none;
  }
}

.moviesGrid_categories_item_block.active {
  display: block;
  cursor: pointer;
  font-size: 18px;
  line-height: 25px;
  color: #0aaee4;
  font-family: Helvetica, Arial, Verdana, Tahoma, sans-serif;
  text-align: left;
  margin-left: 15px;
  margin-top: 10px;
  padding-left: 0;
  position: absolute;
  z-index: 6;
  background: #f5f5f5;
  height: 200px;
  width: 200px;
  overflow-y: scroll;
}

.moviesGrid_categories_item_button.active {
  border-bottom: 4px solid #0aaee4 !important;
}

@media (max-width: 767px) {
  .moviesGrid_categories_item_button {
    font-size: 15px;
    line-height: 18px;
    margin-right: 0;
  }

  .moviesGrid_categories_item_block.active {
    left: 45%;
  }
}
