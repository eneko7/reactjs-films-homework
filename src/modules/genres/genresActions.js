import axios from 'axios';
import { fetchFilmsByGenre } from '../films/filmsActions';
import { PUSH_NAVIGATION_LINK } from '../navlinks/navlinksActions';

export const FETCH_GENRES_REQUEST = 'FETCH_GENRES_REQUEST';
export const FETCH_GENRES_SUCCESS = 'FETCH_GENRES_SUCCESS';
export const FETCH_GENRES_ERROR = 'FETCH_GENRES_ERROR';
export const PUSH_GENRES_LINK = 'PUSH_GENRES_LINK';

export const urlGetGenres = 'https://api.themoviedb.org/3/genre/movie/list?api_key=6a08c0def237c5910708279c9ee78cc5&language=en-US';

export const receiveGenresSuccess = payload => ({
  type: FETCH_GENRES_SUCCESS,
  payload,
});

export const receiveGenresRequest = () => ({
  type: FETCH_GENRES_REQUEST,
});

export const receiveGenresError = payload => ({
  type: FETCH_GENRES_ERROR,
  payload,
});

export function fetchGenres() {
  return (dispatch) => {
    dispatch(receiveGenresRequest());
    axios.get(urlGetGenres)
      .then((response) => {
        dispatch(receiveGenresSuccess(response.data.genres));
      })
      .catch(error => dispatch(receiveGenresError(error.message)));
  };
}
export const pushGenresLink = payload => (dispatch) => {
  dispatch({
    type: PUSH_NAVIGATION_LINK,
    payload: 'genre',
  });
  dispatch({
    type: PUSH_GENRES_LINK,
    payload,
  });
  dispatch(fetchFilmsByGenre(payload));
};
