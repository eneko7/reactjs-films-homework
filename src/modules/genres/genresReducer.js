import {
  FETCH_GENRES_REQUEST,
  FETCH_GENRES_SUCCESS,
  FETCH_GENRES_ERROR,
} from './genresActions';

const initialState = {
  allGenres: [],
  errorGenres: {},
  isFetchingGenres: false,
  isFetchedGenres: false,
  lastGenreID: 0,
};

const genresReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_GENRES_SUCCESS:
      return {
        ...state,
        allGenres: action.payload,
        errorGenres: {},
        isFetchingGenres: false,
        isFetchedGenres: true,
        lastGenreID: 0,
      };
    case FETCH_GENRES_REQUEST:
      return {
        ...state,
        isFetchingGenres: true,
        isFetchedGenres: false,
      };
    case FETCH_GENRES_ERROR:
      return {
        ...state,
        errorGenres: {},
      };
    default:
      return {
        ...state,
      };
  }
};

export default genresReducer;
