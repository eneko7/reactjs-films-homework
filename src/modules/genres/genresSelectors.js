export const getGenres = state => state.genres.allGenres;
export const getGenresId = state => state.genres.lastGenreID;
