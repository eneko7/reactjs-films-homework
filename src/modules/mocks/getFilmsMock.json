{"page":1,"total_results":19767,"total_pages":989,"results":[{"vote_count":1530,"id":335983,"video":false,"vote_average":6.6,"title":"Venom","popularity":389.622,"poster_path":"\/2uNW4WbgBXL25BAbXGLnLqX71Sw.jpg","original_language":"en","original_title":"Venom","genre_ids":[878,28,80,28,27],"backdrop_path":"\/VuukZLgaCrho2Ar8Scl9HtV3yD.jpg","adult":false,"overview":"When Eddie Brock acquires the powers of a symbiote, he will have to release his alter-ego \"Venom\" to save his life.","release_date":"2018-10-03"},{"vote_count":963,"id":332562,"video":false,"vote_average":7.5,"title":"A Star Is Born","popularity":204.707,"poster_path":"\/wrFpXMNBRj2PBiN4Z5kix51XaIZ.jpg","original_language":"en","original_title":"A Star Is Born","genre_ids":[18,10402,10749],"backdrop_path":"\/840rbblaLc4SVxm8gF3DNdJ0YAE.jpg","adult":false,"overview":"Seasoned musician Jackson Maine discovers—and falls in love with—struggling artist Ally. She has just about given up on her dream to make it big as a singer—until Jack coaxes her into the spotlight. But even as Ally's career takes off, the personal side of their relationship is breaking down, as Jack fights an ongoing battle with his own internal demons.","release_date":"2018-10-03"},{"vote_count":44,"id":507569,"video":false,"vote_average":5.9,"title":"The Seven Deadly Sins: Prisoners of the Sky","popularity":184.641,"poster_path":"\/r6pPUVUKU5eIpYj4oEzidk5ZibB.jpg","original_language":"ja","original_title":"劇場版 七つの大罪 天空の囚われ人","genre_ids":[28,12,14,16],"backdrop_path":"\/uKwOX7MtKlAaGeCQe6c4jc1vZpj.jpg","adult":false,"overview":"Traveling in search of the rare ingredient, “sky fish”  Meliodas and Hawk arrive at a palace that floats above the clouds. The people there are busy preparing a ceremony, meant to protect their home from a ferocious beast that awakens once every 3,000 years. But before the ritual is complete, the Six Knights of Black—a Demon Clan army—removes the seal on the beast, threatening the lives of everyone in the Sky Palace.","release_date":"2018-08-18"},{"vote_count":172,"id":424139,"video":false,"vote_average":6.7,"title":"Halloween","popularity":183.991,"poster_path":"\/lNkDYKmrVem1J0aAfCnQlJOCKnT.jpg","original_language":"en","original_title":"Halloween","genre_ids":[27],"backdrop_path":"\/hO1oTBGNxO5fBKVEuWnSpICJH7c.jpg","adult":false,"overview":"Laurie Strode comes to her final confrontation with Michael Myers, the masked figure who has haunted her since she narrowly escaped his killing spree on Halloween night four decades ago.","release_date":"2018-10-18"},{"vote_count":8918,"id":299536,"video":false,"vote_average":8.3,"title":"Avengers: Infinity War","popularity":150.329,"poster_path":"\/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg","original_language":"en","original_title":"Avengers: Infinity War","genre_ids":[12,878,28,14],"backdrop_path":"\/lmZFxXgJE3vgrciwuDib0N8CfQo.jpg","adult":false,"overview":"As the Avengers and their allies have continued to protect the world from threats too large for any one hero to handle, a new danger has emerged from the cosmic shadows: Thanos. A despot of intergalactic infamy, his goal is to collect all six Infinity Stones, artifacts of unimaginable power, and use them to inflict his twisted will on all of reality. Everything the Avengers have fought for has led up to this moment - the fate of Earth and existence itself has never been more uncertain.","release_date":"2018-04-25"},{"vote_count":148,"id":347375,"video":false,"vote_average":5.7,"title":"Mile 22","popularity":145.524,"poster_path":"\/dT1XzjxRDR56xOm4Ph0INV4EmWJ.jpg","original_language":"en","original_title":"Mile 22","genre_ids":[28],"backdrop_path":"\/eTwftrInxzZTSLUkX5hoifOczKJ.jpg","adult":false,"overview":"A CIA field officer and an Indonesian police officer are forced to work together in confronting political corruption. An informant must be moved twenty-two miles to safety.","release_date":"2018-08-16"},{"vote_count":2935,"id":363088,"video":false,"vote_average":7,"title":"Ant-Man and the Wasp","popularity":135.406,"poster_path":"\/rv1AWImgx386ULjcf62VYaW8zSt.jpg","original_language":"en","original_title":"Ant-Man and the Wasp","genre_ids":[28,12,878,10749,35,10751],"backdrop_path":"\/6P3c80EOm7BodndGBUAJHHsHKrp.jpg","adult":false,"overview":"Just when his time under house arrest is about to end, Scott Lang puts again his freedom at risk to help Hope van Dyne and Dr. Hank Pym dive into the quantum realm and try to accomplish, against time and any chance of success, a very dangerous rescue mission.","release_date":"2018-07-04"},{"vote_count":647,"id":346910,"video":false,"vote_average":5.3,"title":"The Predator","popularity":133.473,"poster_path":"\/wMq9kQXTeQCHUZOG4fAe5cAxyUA.jpg","original_language":"en","original_title":"The Predator","genre_ids":[27,878,28,53],"backdrop_path":"\/f4E0ocYeToEuXvczZv6QArrMDJ.jpg","adult":false,"overview":"From the outer reaches of space to the small-town streets of suburbia, the hunt comes home. Now, the universe’s most lethal hunters are stronger, smarter and deadlier than ever before, having genetically upgraded themselves with DNA from other species. When a young boy accidentally triggers their return to Earth, only a ragtag crew of ex-soldiers and a disgruntled science teacher can prevent the end of the human race.","release_date":"2018-09-13"},{"vote_count":1184,"id":439079,"video":false,"vote_average":5.8,"title":"The Nun","popularity":113.535,"poster_path":"\/sFC1ElvoKGdHJIWRpNB3xWJ9lJA.jpg","original_language":"en","original_title":"The Nun","genre_ids":[27,9648,53],"backdrop_path":"\/fgsHxz21B27hOOqQBiw9L6yWcM7.jpg","adult":false,"overview":"When a young nun at a cloistered abbey in Romania takes her own life, a priest with a haunted past and a novitiate on the threshold of her final vows are sent by the Vatican to investigate. Together they uncover the order’s unholy secret. Risking not only their lives but their faith and their very souls, they confront a malevolent force in the form of the same demonic nun that first terrorized audiences in “The Conjuring 2,” as the abbey becomes a horrific battleground between the living and the damned.","release_date":"2018-09-05"},{"vote_count":280,"id":369972,"video":false,"vote_average":7.3,"title":"First Man","popularity":112.81,"poster_path":"\/i91mfvFcPPlaegcbOyjGgiWfZzh.jpg","original_language":"en","original_title":"First Man","genre_ids":[36,18],"backdrop_path":"\/z6ZANhPCcPl1HsKGjY5iZALf65V.jpg","adult":false,"overview":"A look at the life of the astronaut, Neil Armstrong, and the legendary space mission that led him to become the first man to walk on the Moon on July 20, 1969.","release_date":"2018-10-11"},{"vote_count":2987,"id":260513,"video":false,"vote_average":7.6,"title":"Incredibles 2","popularity":105.701,"poster_path":"\/x1txcDXkcM65gl7w20PwYSxAYah.jpg","original_language":"en","original_title":"Incredibles 2","genre_ids":[28,12,16,10751],"backdrop_path":"\/mabuNsGJgRuCTuGqjFkWe1xdu19.jpg","adult":false,"overview":"Elastigirl springs into action to save the day, while Mr. Incredible faces his greatest challenge yet – taking care of the problems of his three children.","release_date":"2018-06-14"},{"vote_count":270,"id":454992,"video":false,"vote_average":6.5,"title":"The Spy Who Dumped Me","popularity":92.441,"poster_path":"\/2lIr27lBdxCpzYDl6WUHzzD6l6H.jpg","original_language":"en","original_title":"The Spy Who Dumped Me","genre_ids":[28,35,12],"backdrop_path":"\/uN6v3Hz4qI2CIqT1Ro4vPgAbub3.jpg","adult":false,"overview":"Audrey and Morgan are best friends who unwittingly become entangled in an international conspiracy when one of the women discovers the boyfriend who dumped her was actually a spy.","release_date":"2018-08-02"},{"vote_count":218,"id":463272,"video":false,"vote_average":6.2,"title":"Johnny English Strikes Again","popularity":89.238,"poster_path":"\/tCBxnZwLiY1BOKw3tH6AxHZdqPh.jpg","original_language":"en","original_title":"Johnny English Strikes Again","genre_ids":[12,10751,28,35],"backdrop_path":"\/yCOLqh5MOGyYdo58Ap0aWvKop9h.jpg","adult":false,"overview":"Disaster strikes when a criminal mastermind reveals the identities of all active undercover agents in Britain. The secret service can now rely on only one man—Johnny English. Currently teaching at a minor prep school, Johnny springs back into action to find the mysterious hacker. For this mission to succeed, he’ll need all of his skills—what few he has—as the man with yesterday’s analogue methods faces off against tomorrow’s digital technology.","release_date":"2018-09-13"},{"vote_count":8664,"id":284054,"video":false,"vote_average":7.3,"title":"Black Panther","popularity":84.957,"poster_path":"\/uxzzxijgPIY7slzFvMotPv8wjKA.jpg","original_language":"en","original_title":"Black Panther","genre_ids":[28,12,14,878],"backdrop_path":"\/b6ZJZHUdMEFECvGiDpJjlfUWela.jpg","adult":false,"overview":"King T'Challa returns home from America to the reclusive, technologically advanced African nation of Wakanda to serve as his country's new leader. However, T'Challa soon finds that he is challenged for the throne by factions within his own country as well as without. Using powers reserved to Wakandan kings, T'Challa assumes the Black Panther mantel to join with girlfriend Nakia, the queen-mother, his princess-kid sister, members of the Dora Milaje (the Wakandan 'special forces') and an American secret agent, to prevent Wakanda from being dragged into a world war.","release_date":"2018-02-13"},{"vote_count":695,"id":487558,"video":false,"vote_average":7.5,"title":"BlacKkKlansman","popularity":74.27,"poster_path":"\/bT5WuAsjDJYQv2vXbWGDQTmjKav.jpg","original_language":"en","original_title":"BlacKkKlansman","genre_ids":[80,18,35,36],"backdrop_path":"\/gMVdhfQ7q9DFHhDkehrququjGPd.jpg","adult":false,"overview":"Ron Stallworth, an African-American police officer from Colorado, successfully manages to infiltrate the local Ku Klux Klan and become the head of the local chapter.","release_date":"2018-08-09"},{"vote_count":4035,"id":351286,"video":false,"vote_average":6.5,"title":"Jurassic World: Fallen Kingdom","popularity":73.868,"poster_path":"\/c9XxwwhPHdaImA2f1WEfEsbhaFB.jpg","original_language":"en","original_title":"Jurassic World: Fallen Kingdom","genre_ids":[28,12,878],"backdrop_path":"\/3s9O5af2xWKWR5JzP2iJZpZeQQg.jpg","adult":false,"overview":"Three years after the demise of Jurassic World, a volcanic eruption threatens the remaining dinosaurs on the isla Nublar, so Claire Dearing, the former park manager, recruits Owen Grady to help prevent the extinction of the dinosaurs once again.","release_date":"2018-06-06"},{"vote_count":5725,"id":383498,"video":false,"vote_average":7.5,"title":"Deadpool 2","popularity":73.355,"poster_path":"\/to0spRl1CMDvyUbOnbb4fTk3VAd.jpg","original_language":"en","original_title":"Deadpool 2","genre_ids":[28,35,878],"backdrop_path":"\/3P52oz9HPQWxcwHOwxtyrVV1LKi.jpg","adult":false,"overview":"Wisecracking mercenary Deadpool battles the evil and powerful Cable and other bad guys to save a boy's life.","release_date":"2018-05-15"},{"vote_count":1268,"id":345940,"video":false,"vote_average":6,"title":"The Meg","popularity":66.424,"poster_path":"\/eyWICPcxOuTcDDDbTMOZawoOn8d.jpg","original_language":"en","original_title":"The Meg","genre_ids":[28,878,53],"backdrop_path":"\/rH79sB6Nkx4cMW3JzsUy7wK0rhX.jpg","adult":false,"overview":"A deep sea submersible pilot revisits his past fears in the Mariana Trench, and accidentally unleashes the seventy foot ancestor of the Great White Shark believed to be extinct.","release_date":"2018-08-09"},{"vote_count":11413,"id":671,"video":false,"vote_average":7.8,"title":"Harry Potter and the Philosopher's Stone","popularity":65.451,"poster_path":"\/dCtFvscYcXQKTNvyyaQr2g2UacJ.jpg","original_language":"en","original_title":"Harry Potter and the Philosopher's Stone","genre_ids":[12,14,10751],"backdrop_path":"\/dug34qm7BSnC1AVwtuQsGDFxxbG.jpg","adult":false,"overview":"Harry Potter has lived under the stairs at his aunt and uncle's house his whole life. But on his 11th birthday, he learns he's a powerful wizard -- with a place waiting for him at the Hogwarts School of Witchcraft and Wizardry. As he learns to harness his newfound powers with the help of the school's kindly headmaster, Harry uncovers the truth about his parents' deaths -- and about the villain who's to blame.","release_date":"2001-11-16"},{"vote_count":41,"id":442062,"video":false,"vote_average":5.9,"title":"Goosebumps 2: Haunted Halloween","popularity":65.186,"poster_path":"\/8bcpki9GfXdXj9esFpPtlate8v0.jpg","original_language":"en","original_title":"Goosebumps 2: Haunted Halloween","genre_ids":[12,35,14,27,10751],"backdrop_path":"\/9xYg9ScjGaO2HrLfOfYZirTrB3o.jpg","adult":false,"overview":"Two boys face an onslaught from witches, monsters, ghouls and a talking dummy after they discover a mysterious book by author R. L. Stine.","release_date":"2018-10-11"}]}