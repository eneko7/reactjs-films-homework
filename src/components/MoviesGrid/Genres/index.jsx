import Genres from './GenresContainer';

export default Genres;
