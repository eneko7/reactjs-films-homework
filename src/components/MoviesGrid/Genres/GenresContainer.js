import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import Genres from './Genres';
import { fetchGenres } from '../../../modules/genres/genresActions';
import { getGenres, getGenresId } from '../../../modules/genres/genresSelectors';
import { fetchFilmsByGenre } from '../../../modules/films/filmsActions';

const mapStateToProps = state => ({
  genres: getGenres(state),
  genresLastId: getGenresId(state),
});

const mapDispatchToProps = {
  fetchGenres,
  fetchFilmsByGenre,
};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Genres));
