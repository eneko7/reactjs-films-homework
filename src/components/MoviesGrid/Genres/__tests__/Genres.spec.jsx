import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import moxios from 'moxios';
import { Provider } from 'react-redux';
import getGenresMock from '../../../../modules/mocks/getGenresMock';
import getFilmsMock from '../../../../modules/mocks/getFilmsMock';
import * as actionsGenres from '../../../../modules/genres/genresActions';
import Genres from '../index';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  films: {
    allFilms: getFilmsMock.results,
    isFetchingFilms: false,
  },
  navlinks: {
    clickedLink: 'Trending',
  },
  genres: {
    allGenres: [
      {
        id: 1,
        name: 'adventure',
      },
      {
        id: 2,
        name: 'action',
      },
      {
        id: 3,
        name: 'drama',
      },
    ],
    errorGenres: {},
    isFetchingGenres: false,
    isFetchedGenres: false,
    lastGenreID: 1,
  },
});

const shallow = new ShallowRenderer();
describe('GenresNavLink Snapshot', () => {
  test('renders', () => {
    const component = shallow.render(
      <Provider store={store}>
        <Genres title="Genres" activeClass="undefined" fetchChange={() => ('Hello')} fetchFilmsByGenre={() => ('Hello')} />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
  test('renders', () => {
    const component = shallow.render(
      <Provider store={store}>
        <Genres title="Genres" activeClass="active" fetchChange={() => ('Hello')} fetchFilmsByGenre={() => ('Hello')} />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});
jest.mock('../../../../modules/films/filmsActions',
  () => ({
    fetchGenres: jest.fn(() => ({ type: 'FETCH_GENRES_REQUEST' })),
    fetchFilmsByGenre: jest.fn(() => ({ type: 'FETCH_FILMS_REQUEST' })),
  }));
describe('Genres logic', () => {
  it('renders genres dropdown open', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Genres title="Genres" activeClass="active" fetchChange={() => ('Hello')} fetchFilmsByGenre={() => ('Hello')} />
      </Provider>,
    );
    component.root.findByProps({ className: 'moviesGrid_categories_item_button' }).props.onClick();
    expect(component).toMatchSnapshot();
  });

  it('renders genres dropdown close', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Genres title="Genres" activeClass="undefined" fetchChange={() => ('Hello')} fetchFilmsByGenre={() => ('Hello')} />
      </Provider>,
    );
    component.root.findByProps({ className: 'moviesGrid_categories_item_button' }).props.onClick();
    component.root.findByProps({ className: 'moviesGrid_categories_item_button' }).props.onClick();
    expect(component).toMatchSnapshot();
  });

  it('renders genres dropdown chose', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Genres title="Genres" activeClass="active" fetchChange={() => ('Hello')} fetchFilmsByGenre={() => ('Hello')} />
      </Provider>,
    );
    const { root } = component;
    root.findByProps({ className: 'moviesGrid_categories_item_button' }).props.onClick();
    root.findByProps({ id: 'genres_list' }).children[0].props.onClick();
    expect(component).toMatchSnapshot();
  });

  it('renders genres dropdown chose', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <Genres title="Genres" activeClass="undefined" fetchChange={() => ('Hello')} fetchFilmsByGenre={() => ('Hello')} />
      </Provider>,
    );
    const { root } = component;
    root.findByProps({ className: 'moviesGrid_categories_item_button' }).props.onClick();
    root.findByProps({ id: 'genres_list' }).children[0].props.onClick();
    expect(component).toMatchSnapshot();
  });

  beforeEach(() => {
    moxios.install();
    store.clearActions();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('creates GET_GENRES_SUCCESS after successfuly fetching genres', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: getGenresMock,
      });
    });
    const expectedActions = [
      { type: actionsGenres.FETCH_GENRES_REQUEST },
      { type: actionsGenres.FETCH_GENRES_SUCCESS, payload: getGenresMock.genres },
    ];
    return store.dispatch(actionsGenres.fetchGenres()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
