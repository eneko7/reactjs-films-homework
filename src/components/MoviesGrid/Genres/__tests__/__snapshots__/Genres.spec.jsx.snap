// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genres logic renders genres dropdown chose 1`] = `
<div>
  <button
    className="moviesGrid_categories_item_button"
    data-filter="Genres"
    id="Genres"
    onClick={[Function]}
    style={
      Object {
        "borderBottom": "4px solid",
      }
    }
    type="button"
  >
    <span>
      adventure
    </span>
    <span
      className=" moviesGrid_categories_item_span_arrow"
    />
  </button>
  <div
    className=" moviesGrid_categories_item_block"
    id="genres_list"
  >
    <button
      className="moviesGrid_categories_item_block_button"
      elem="adventure"
      id="adventure"
      onClick={[Function]}
      type="button"
    >
      adventure
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="action"
      id="action"
      onClick={[Function]}
      type="button"
    >
      action
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="drama"
      id="drama"
      onClick={[Function]}
      type="button"
    >
      drama
    </button>
  </div>
</div>
`;

exports[`Genres logic renders genres dropdown chose 2`] = `
<div>
  <button
    className="moviesGrid_categories_item_button"
    data-filter="Genres"
    id="Genres"
    onClick={[Function]}
    style={
      Object {
        "borderBottom": "none",
      }
    }
    type="button"
  >
    <span>
      adventure
    </span>
    <span
      className=" moviesGrid_categories_item_span_arrow"
    />
  </button>
  <div
    className=" moviesGrid_categories_item_block"
    id="genres_list"
  >
    <button
      className="moviesGrid_categories_item_block_button"
      elem="adventure"
      id="adventure"
      onClick={[Function]}
      type="button"
    >
      adventure
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="action"
      id="action"
      onClick={[Function]}
      type="button"
    >
      action
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="drama"
      id="drama"
      onClick={[Function]}
      type="button"
    >
      drama
    </button>
  </div>
</div>
`;

exports[`Genres logic renders genres dropdown close 1`] = `
<div>
  <button
    className="moviesGrid_categories_item_button"
    data-filter="Genres"
    id="Genres"
    onClick={[Function]}
    style={
      Object {
        "borderBottom": "none",
      }
    }
    type="button"
  >
    <span>
      Genres
    </span>
    <span
      className=" moviesGrid_categories_item_span_arrow"
    />
  </button>
  <div
    className=" moviesGrid_categories_item_block"
    id="genres_list"
  >
    <button
      className="moviesGrid_categories_item_block_button"
      elem="adventure"
      id="adventure"
      onClick={[Function]}
      type="button"
    >
      adventure
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="action"
      id="action"
      onClick={[Function]}
      type="button"
    >
      action
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="drama"
      id="drama"
      onClick={[Function]}
      type="button"
    >
      drama
    </button>
  </div>
</div>
`;

exports[`Genres logic renders genres dropdown open 1`] = `
<div>
  <button
    className="moviesGrid_categories_item_button"
    data-filter="Genres"
    id="Genres"
    onClick={[Function]}
    style={
      Object {
        "borderBottom": "4px solid",
      }
    }
    type="button"
  >
    <span>
      Genres
    </span>
    <span
      className="active moviesGrid_categories_item_span_arrow"
    />
  </button>
  <div
    className="active moviesGrid_categories_item_block"
    id="genres_list"
  >
    <button
      className="moviesGrid_categories_item_block_button"
      elem="adventure"
      id="adventure"
      onClick={[Function]}
      type="button"
    >
      adventure
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="action"
      id="action"
      onClick={[Function]}
      type="button"
    >
      action
    </button>
    <button
      className="moviesGrid_categories_item_block_button"
      elem="drama"
      id="drama"
      onClick={[Function]}
      type="button"
    >
      drama
    </button>
  </div>
</div>
`;

exports[`GenresNavLink Snapshot renders 1`] = `
<Connect(Genres)
  activeClass="undefined"
  fetchChange={[Function]}
  fetchFilmsByGenre={[Function]}
  title="Genres"
/>
`;

exports[`GenresNavLink Snapshot renders 2`] = `
<Connect(Genres)
  activeClass="active"
  fetchChange={[Function]}
  fetchFilmsByGenre={[Function]}
  title="Genres"
/>
`;
